api: generate markdown documentation for gRPC APIs

rolinh opened the pull request:
See commits for details.

rolinh committed:
images/builder: bump protoc to v3.20.1

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
images/builder: bump protoc-gen-go-grpc to v1.2.0

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
images/builder: add protoc-gen-doc plugin

The protoc-gen-doc plugin allows to generate documentation from proto
files as Markdown, HTML, DocBook and JSON.

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
images: update cilium-builder

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
api: generate proto API documentation as markdown

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

nathanjsweet reviewed (approved):
Nice! Is this just for source code users, or do we plan on importing this stuff into the main docs somehow?

tklauser reviewed (approved):
🚀 nice!

qmonnet commented on Documentation/Makefile:
```diff
@@ -88,7 +88,7 @@ $(HELM_VALUES): FORCE
 	$(QUIET)$(RM) -- $(TMP_FILE_1) $(TMP_FILE_2)
 
-epub latex html: builder-image copy-api update-helm-values
+epub latex html: clean builder-image copy-api update-helm-values
```
Why is the clean-up required? Did you experience any issue?

rolinh replied:
Yes, this target is not idempotent because of the `copy-api` step. The 1st time it's invoked,  it copies  `../api` to `_api` and the 2nd time, as the `_api` directory already exists, it gets copied to `_api/api`. I guess a better fix would be to `rm -rf _api` at the `copy-api` target.

qmonnet replied:
Even better if we could copy correctly. I think we could pass `-T` to `cp` to avoid copying into `_api/`?

```
copy-api:
	@$(ECHO_GEN)_api
	$(QUIET)cp -T -r ../api _api
```

qmonnet reviewed (approved):
Looks good to me, let's make sure the new docs-builder image works fine before merging

rolinh committed:
images/builder: bump protoc-gen-go to v1.28.0

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
ci: run doc dependencies before checking doc

The CI step runs `check-build.sh` without using the `make html` target.
This means that the `copy-api` target, which is a pre-requisite for
`make html` isn't run. However, running the whole `make html` target
would be too costly as it would run the slow builder-image and update
helm values targets. Therefore, we simply ensure to run `make copy-api`
before running `check-build.sh`.

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

pchaigno commented on .github/workflows/documentation.yaml:
```diff
@@ -53,6 +53,8 @@ jobs:
+        make -C Documentation copy-api # necessary for check-build.sh
       - uses: docker://cilium/docs-builder:2022-04-08@sha256:44f59d844a7e1162080e6fdd62c72a16fc268c893c6cb81e1fbd31cfa189e691
         with:
           entrypoint: ./Documentation/check-build.sh
```
Would it be worth having a make target just for `check-build.sh` with `copy-api` explicitly defined as a dependency?

qmonnet replied:
This would avoid adding an additional step in the description of the GitHub Action, but would require a step (target) in the Makefile which is not otherwise needed, given that all targets calling check-build.sh already have a dependency on `copy-api`. So I'm good with the current version for my part. I don't mind another target but I'm not convinced it's any simpler.

rolinh committed:
docs: fix check-build.sh exit status

The `check-build.sh` script should only return with exit 1 if errors
have actually been found, even for the non-spelling case.

Fixes: f98307558301044c44fe69a81eea2cd244e18752

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

qmonnet reviewed (approved):
Thanks!
